File: python/bc-components/src/bc_components/symmetric/_encrypted_message.py
"""Encrypted message using ChaCha20-Poly1305 AEAD.

Contains ciphertext, nonce, authentication tag, and optional additional
authenticated data (AAD).
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from bc_tags import TAG_ENCRYPTED, tags_for_values
from dcbor import CBOR, Tag

from .._digest import Digest
from .._error import BCComponentsError
from .._nonce import Nonce
from ._authentication_tag import AuthenticationTag

if TYPE_CHECKING:
    from bc_ur import UR


class EncryptedMessage:
    """A ChaCha20-Poly1305 encrypted message with optional AAD.

    CDDL::

        EncryptedMessage =
            #6.40002([ ciphertext: bstr, nonce: bstr, auth: bstr, ? aad: bstr ])
    """

    __slots__ = ("_ciphertext", "_aad", "_nonce", "_auth")

    def __init__(
        self,
        ciphertext: bytes | bytearray,
        aad: bytes | bytearray,
        nonce: Nonce,
        auth: AuthenticationTag,
    ) -> None:
        self._ciphertext = bytes(ciphertext)
        self._aad = bytes(aad)
        self._nonce = nonce
        self._auth = auth

    # --- Construction ---

    @staticmethod
    def new(
        ciphertext: bytes | bytearray,
        aad: bytes | bytearray,
        nonce: Nonce,
        auth: AuthenticationTag,
    ) -> EncryptedMessage:
        """Restore an EncryptedMessage from its component parts."""
        return EncryptedMessage(ciphertext, aad, nonce, auth)

    # --- Accessors ---

    @property
    def ciphertext(self) -> bytes:
        """The encrypted data."""
        return self._ciphertext

    @property
    def aad(self) -> bytes:
        """Additional authenticated data (may be empty)."""
        return self._aad

    @property
    def nonce(self) -> Nonce:
        """The 12-byte nonce used for encryption."""
        return self._nonce

    @property
    def authentication_tag(self) -> AuthenticationTag:
        """The 16-byte Poly1305 authentication tag."""
        return self._auth

    # --- AAD helpers ---

    def aad_cbor(self) -> CBOR | None:
        """Parse the AAD as CBOR, or return ``None``."""
        if not self._aad:
            return None
        try:
            return CBOR.from_data(self._aad)
        except Exception:
            return None

    def aad_digest(self) -> Digest | None:
        """Parse the AAD as a tagged CBOR ``Digest``, or return ``None``."""
        cbor = self.aad_cbor()
        if cbor is None:
            return None
        try:
            return Digest.from_tagged_cbor(cbor)
        except Exception:
            return None

    @property
    def has_digest(self) -> bool:
        """``True`` if the AAD contains a valid tagged Digest."""
        return self.aad_digest() is not None

    # --- DigestProvider ---

    def digest(self) -> Digest:
        """Return the digest stored in the AAD (raises if absent)."""
        d = self.aad_digest()
        if d is not None:
            return d
        raise BCComponentsError.invalid_data(
            "EncryptedMessage", "AAD does not contain a valid digest"
        )

    # --- CBOR ---

    @staticmethod
    def cbor_tags() -> list[Tag]:
        return tags_for_values([TAG_ENCRYPTED])

    def untagged_cbor(self) -> CBOR:
        items: list[CBOR] = [
            CBOR.from_bytes(self._ciphertext),
            CBOR.from_bytes(self._nonce.data),
            CBOR.from_bytes(self._auth.data),
        ]
        if self._aad:
            items.append(CBOR.from_bytes(self._aad))
        return CBOR.from_array(items)

    def tagged_cbor(self) -> CBOR:
        tags = self.cbor_tags()
        return CBOR.from_tagged_value(tags[0], self.untagged_cbor())

    def tagged_cbor_data(self) -> bytes:
        return self.tagged_cbor().to_cbor_data()

    @classmethod
    def from_untagged_cbor(cls, cbor: CBOR) -> EncryptedMessage:
        elements = cbor.try_array()
        if len(elements) < 3:
            raise BCComponentsError.invalid_data(
                "EncryptedMessage", "must have at least 3 elements"
            )
        ciphertext = elements[0].try_byte_string()
        nonce_data = elements[1].try_byte_string()
        nonce = Nonce.from_data(nonce_data)
        auth_data = elements[2].try_byte_string()
        auth = AuthenticationTag.from_data(auth_data)
        aad = elements[3].try_byte_string() if len(elements) > 3 else b""
        return cls(ciphertext, aad, nonce, auth)

    @classmethod
    def from_tagged_cbor(cls, cbor: CBOR) -> EncryptedMessage:
        tags = cls.cbor_tags()
        item = cbor.try_expected_tagged_value(tags[0])
        return cls.from_untagged_cbor(item)

    @classmethod
    def from_tagged_cbor_data(cls, data: bytes) -> EncryptedMessage:
        cbor = CBOR.from_data(data)
        return cls.from_tagged_cbor(cbor)

    # --- UR ---

    def to_ur(self) -> UR:
        from bc_ur import to_ur
        return to_ur(self)

    def ur_string(self) -> str:
        from bc_ur import to_ur_string
        return to_ur_string(self)

    @classmethod
    def from_ur(cls, ur: UR) -> EncryptedMessage:
        from bc_ur import from_ur
        return from_ur(cls, ur)

    @classmethod
    def from_ur_string(cls, ur_string: str) -> EncryptedMessage:
        from bc_ur import from_ur_string
        return from_ur_string(cls, ur_string)

    # --- Dunder methods ---

    def __eq__(self, other: object) -> bool:
        if isinstance(other, EncryptedMessage):
            return (
                self._ciphertext == other._ciphertext
                and self._aad == other._aad
                and self._nonce == other._nonce
                and self._auth == other._auth
            )
        return NotImplemented

    def __hash__(self) -> int:
        return hash((self._ciphertext, self._aad, self._nonce, self._auth))

    def __repr__(self) -> str:
        return (
            f"EncryptedMessage("
            f"ciphertext={self._ciphertext.hex()}, "
            f"aad={self._aad.hex()}, "
            f"nonce={self._nonce!r}, "
            f"auth={self._auth!r})"
        )

    def __str__(self) -> str:
        return f"EncryptedMessage(ciphertext={self._ciphertext.hex()})"

File: typescript/bc-components/src/symmetric/encrypted-message.ts
import {
    type Cbor,
    decodeCbor,
    cbor as toCbor,
    createTag,
    toByteString,
} from '@bc/dcbor';
import { TAG_ENCRYPTED } from '@bc/tags';
import type { Tag } from '@bc/dcbor';

import { decodeTaggedCbor, defaultTaggedCbor, defaultTaggedCborData, fromTaggedUrString, toUrString } from '../cbor-ur.js';
import { Digest } from '../digest.js';
import { type DigestProvider } from '../digest-provider.js';
import { bytesEqual, hexEncode } from '../utils.js';
import { AuthenticationTag } from './authentication-tag.js';
import { Nonce } from '../nonce.js';

const ENCRYPTED_TAG: Tag = createTag(TAG_ENCRYPTED, 'encrypted');

export class EncryptedMessage implements DigestProvider {
    readonly #ciphertext: Uint8Array;
    readonly #aad: Uint8Array;
    readonly #nonce: Nonce;
    readonly #auth: AuthenticationTag;

    constructor(
        ciphertext: Uint8Array,
        aad: Uint8Array,
        nonce: Nonce,
        auth: AuthenticationTag,
    ) {
        this.#ciphertext = new Uint8Array(ciphertext);
        this.#aad = new Uint8Array(aad);
        this.#nonce = nonce;
        this.#auth = auth;
    }

    static new(
        ciphertext: Uint8Array,
        aad: Uint8Array,
        nonce: Nonce,
        auth: AuthenticationTag,
    ): EncryptedMessage {
        return new EncryptedMessage(ciphertext, aad, nonce, auth);
    }

    get ciphertext(): Uint8Array {
        return new Uint8Array(this.#ciphertext);
    }

    get aad(): Uint8Array {
        return new Uint8Array(this.#aad);
    }

    get nonce(): Nonce {
        return this.#nonce;
    }

    get authenticationTag(): AuthenticationTag {
        return this.#auth;
    }

    aadCbor(): Cbor | undefined {
        if (this.#aad.length === 0) {
            return undefined;
        }
        try {
            return decodeCbor(this.#aad);
        } catch {
            return undefined;
        }
    }

    aadDigest(): Digest | undefined {
        const value = this.aadCbor();
        if (value === undefined) {
            return undefined;
        }
        try {
            return Digest.fromCbor(value);
        } catch {
            return undefined;
        }
    }

    hasDigest(): boolean {
        return this.aadDigest() !== undefined;
    }

    digest(): Digest {
        return this.aadDigest() ?? Digest.fromImage(this.#ciphertext);
    }

    equals(other: unknown): boolean {
        return (
            other instanceof EncryptedMessage &&
            bytesEqual(this.#ciphertext, other.#ciphertext) &&
            bytesEqual(this.#aad, other.#aad) &&
            this.#nonce.equals(other.#nonce) &&
            this.#auth.equals(other.#auth)
        );
    }

    cborTags(): Tag[] {
        return [ENCRYPTED_TAG];
    }

    untaggedCbor(): Cbor {
        const arr: Cbor[] = [
            toByteString(this.#ciphertext),
            toByteString(this.#nonce.data),
            toByteString(this.#auth.data),
        ];
        if (this.#aad.length > 0) {
            arr.push(toByteString(this.#aad));
        }
        return toCbor(arr);
    }

    taggedCbor(): Cbor {
        return defaultTaggedCbor(this);
    }

    taggedCborData(): Uint8Array {
        return defaultTaggedCborData(this);
    }

    cbor(): Cbor {
        return this.taggedCbor();
    }

    urString(): string {
        return toUrString(this);
    }

    static fromURString(value: string): EncryptedMessage {
        return fromTaggedUrString(EncryptedMessage, value);
    }

    static cborTags(): Tag[] {
        return [ENCRYPTED_TAG];
    }

    static fromUntaggedCbor(cbor: Cbor): EncryptedMessage {
        const arr = cbor.toArray();
        if (arr.length < 3) {
            throw new Error('EncryptedMessage must have at least 3 elements');
        }
        const ciphertext = arr[0]!.toByteString();
        const nonce = Nonce.fromData(arr[1]!.toByteString());
        const auth = AuthenticationTag.fromData(arr[2]!.toByteString());
        const aad = arr[3]?.toByteString() ?? new Uint8Array(0);
        return new EncryptedMessage(ciphertext, aad, nonce, auth);
    }

    static fromCbor(cbor: Cbor): EncryptedMessage {
        return decodeTaggedCbor(EncryptedMessage, cbor);
    }

    toString(): string {
        return `EncryptedMessage(ciphertext=${hexEncode(this.#ciphertext)})`;
    }
}
